### ISSUE TITLE
npm Compatibility

### ISSUE THREAD

**Issue by jmwright** [7BIndustries/sliderule-cli#1]
@m30-jrs I can't find a way to make npm install modules to any other directory than `node_modules`. What this means is that we either have to conform to their directory convention so that npm can be used manually, or we keep the `components` directory name and break manual compatibility. Finding an npm library for Python or Rust that also handles git repos rather than the npm repository was shaping up to be a challenge too, so I'm kind of resigned to blazing my own trail just using npm as inspiration right now.

Thoughts?

**Comment by jmwright** [comment]
If we break form using npm, I think that `bom_data.yaml` and `package.json` could maybe be combined and `bom_data.yaml` could be the source for BoM rendering and package/dependency management. I noticed that not all the components in `bom_data.yaml` are referenced in `package.json`. Maybe that's why they need to be separate files, but if everything is a component, shouldn't everything be referenced in both files, causing a lot of overlap?

**Comment by jmwright** [comment]
> Something is only a dependency if it is externally sourced (otherwise it is just a nested part of the current project's source).

It seems like the local, nested parts are still a dependency in the sense that if they're removed, the project is not whole, and may not work at all.

**Comment by jmwright** [comment]
Another thought: In general, `node_modules` isn't included in commits (it's usually in .gitignore), but there are exceptions and arguments against that. Some people say that including it ensures that if a package disappears out of the npm registry, you still have a copy so that your app doesn't break on `npm install`. The same could be said for Sliderule projects. If a remote component disappears because a user deletes the repo or takes a server down, keeping a copy of `node_modules` in your repo wastes space, but ensures that your project can't be broken by disappearing remote components. I'm not necessarily advocating for keeping `node_modules` in a component repo, but it's a decision that should be intentional. 

It's easy enough for a user to change through .gitignore, and a Sliderule CLI/UI could offer a setting for keeping cached versions of remote components, which could be disabled by default. That might be the best route.

**Comment by jmwright** [comment]
The CLI now has the ability to add/create local components, which don't have any remote (git) setup done and don't have a package.json. The next step will be implementing npm handling so that remote components can be added.

The plan right now is to use npm if it's installed on the system. Otherwise the CLI will fall back to a simpler internal implementation.

**Comment by jmwright** [comment]
@m30-jrs I now have remote component installation working, assuming npm is already installed on the system. Syntactically, npm seems to be run the same way on MacOS, Windows and Linux, but I've only tested Linux so far.

I think this marks a milestone in that the CLI now has the bare minimum feature set to create and manage a project and its components. The next thing I'll work on is a command to refactor a local component to be remote since that's fairly low hanging fruit and is an important, although not immediately needed, feature.

If you are interested, let me know which OS you'll be using the CLI on and I'll build a binary package for that first and make it available for download.

**Comment by jmwright** [comment]
Check that. I forgot about the upload and update commands. Those really need to be in place for an MVP.

**Comment by m30-jrs** [comment]
In reference to storing node_modules in project repos...
> It's easy enough for a user to change through .gitignore, and a Sliderule CLI/UI could offer a setting for keeping cached versions of remote components, which could be disabled by default. That might be the best route.

👍 I think using a setting to control this behavior is the best way to go.  It gives us the opportunity to help the user think about how they want to work (hopefully getting projects to make an informed and intentional decision).  We should discuss the default value for this setting, but I think this is the way to go.

I will be running on Win 10 for the time being, so if you can make a release for that environment I can do some testing/experimenting next week.

> Check that. I forgot about the upload and update commands. Those really need to be in place for an MVP.

Yup, we will definitely need those.  And the update command needs to account for pulling current project down from repo (git pull) as well as syncing data for components from remote repos (npm update).  Uploading is probably just about the commit cycle, so that should be simpler.

We should also discuss how much if any built-in capabilities there should be for managing component source for external components you manage (think ones that have been pulled into their own repo by a refactor command).  I had been writing a lot about helping the user a great deal where this is concerned (creating component repos, checking them out, npm installing them all in one command).  That might be a feature that is better left to layers above the CLI, letting the CLI focus on the structure of a single component repo and its dependencies/contained components.  

Not sure if any of that makes sense....  I need to sit down and do some actual writing and thinking about all of our recent updates and I think it shows...

**Comment by m30-jrs** [comment]
So, I should have looked at the commands before I posted my last message.  I think we have talked in the past about using terminology for interacting with projects that is decoupled from any specific repo implementation and is less code oriented.  And I can see some of that coming out in your current command set.  Here is some feedback along those lines.

_Note: I would make all generation operations independent from project management operations.  So, generating BoMs, docs, etc should be in their own commands and not called from project management commands._

**MVP for Project Management (CRUD for Components)**
* Create-like actions - note, these are really the same operation (mkdir, set up component data structure, etc), just with different destinations
  * Create a new "top-level" component (destination: `./component_name`)
  * Create a new component within the current component's source (destination: `./components/component_name`) 
* Read-like actions
  * Add dependency (aka, read external repo source into component dependency data)
  * Download - the question is how fine of grain control do we want to provide?  I can imagine wanting to be able to download just the current working directory and below (say I am working on a sub-component and I don't want to cause a download across the parent and all sibling components).  And, I can imagine wanting to pull down just updates to dependencies but not data from my repo (for example when I am checking out if updates from dependencies break any of my in development source).  Of course, I can also see wanting to simply download any changes in my repo's source and any required changes to dependencies.  I probably don't want to download just from repo and ignore changes from dependencies triggered by downloading from my repo.
    * download command executes based on current working directory (so can download from entire project by running download all from root component)
    * download all (default, so can leave off all) - downloads from component repo then downloads dependencies recursively from current working directory
    * download dependencies is for updating dependencies in current working directly only (no recursively)
  * What can we do with the `clone` use case?  Currently that is a very git oriented term.  It is like downloading, but the other examples all reference source that we already know the remote location for.
* Update-like actions
  * Upload - upload to repo from current working directory
  * Refactor - basically as written
* Delete-like actions
  * Remove - as written

For transparency (and because it is getting late here), one of my goals was to standardize on download for `git pull` / `npm update` activities and upload for `git push` activities.  These sound generic enough that I think we will stay away from repo terminology related issues.  Users will still have to be in the habit of performing a download, verify (often by running generators and reviewing them), upload cycle when they are ready to share their local work.

**TL;DR**
Here are some proposed revisions to the cli commands.
* create <component_name> - if CWD defines a sliderule component creates component at ./components/component_name, else creates component at ./component_name
* add <component_url> - CWD must define a sliderule component, adds component as a dependency
* download {all | dependencies | component_url}
  * download all - CWD must define a sliderule component, does equivalent of `git pull` followed by `npm update` for CWD and recursively through sub-components
  * download dependencies - CWD must define a sliderule component, does the equivalent of `npm update` in CWD only (no recursion?  Might need optional recursion?)
  * download <component_url> - CWD must **not** define a sliderule component, does the equivalent of a `git clone <component_url>`
* refactor <component_name> - CWD must define a sliderule component, component_name must be a directory under ./components; performs refactor operation on component_name (need to explore what happens when we are in a sub-component and we call this command)
* upload - CWD must define a sliderule component, performs the equivalent of `git commit` followed by `git push`.  Need to address `git add` like steps to identify what files are part of this upload and how to manage differences between other repo tools
* remove <component_name> - CWD must define a sliderule component, if component name is a dependency, removes component_name from dependency list and run equivalent of `npm prune`; if component name is present in ./components then rmdir ./components/component_name at the repo level

**Comment by jmwright** [comment]
The `sliderule-cli component [add|create|etc]` model was what I originally used, but then I simplified it when thinking about the intended audience. I spend most of my time writing code, and it still took me a long time to get used to `git remote add origin` because it's like Yoda-speak compared to the way we think and speak normally. My first thought would always be `git add remote` because that's how I would say it. I don't say "I'm going to remote add", I say "I'm going to add a remote". So I simplified the commands to (for example) `add_component` instead of using a command and a sub-command: `add component`. I'm not saying I'm obstinately opposed to doing the git style command ordering, I just want to make sure we've thought the implications through.  

> We should also discuss how much if any built-in capabilities there should be for managing component source for external components you manage (think ones that have been pulled into their own repo by a refactor command).

I'm not sure how that one will work since I don't think doing `npm install` brings the git directory structure with the module. If I understand you correctly, you're talking about a user making changes to one of their own remote components, and then pushing it back to the remote repo? Originally the CLI grabbed external components using git instead of npm, but dependency management has to be manual if we do that. Maybe we need to add the concepts of "owned" components in addition to local/remote? Local would be assumed to be owned, but remote components could be owned or not owned. If a remote component is owned, we could try to figure out how to use npm to install the component, but also checkout the git structure too. We might have to roll our own solution on that one. This also reminds me of the question of how to handle merge conflicts in the CLI. There was an OSHW group that Mach 30 worked with years ago that complained that GitHub put extra junk in their files that they didn't know how to fix, so they stopped using GitHub to host their files. I'm 99% sure that what they were experiencing was git's handling of merge conflicts. Non-coders can't handle their files getting magically mangled, and when you've got a team managing a bunch of components, there are bound to be conflicts. This topic probably deserves its own issue though.

> I would make all generation operations independent from project management operations.

If we do that, we should probably have pseudo-commands that exist just to run several commands at once to save users from a lot of tedium. For example, `sliderule-cli render` might run BoM export along with CadQuery->STEP export and any other processors/renderers, and put rendered files in `dist`.

> Download - the question is how fine of grain control do we want to provide?

I think what you've laid out with separating pulling the project changes from git vs pulling the remote dependency components is fine, and is something we'd need eventually anyway. I worry about limiting the levels of parent-child relationships, because it seems like you could end up with broken remote components because they're missing a critical piece somewhere down the line.

> What can we do with the clone use case? Currently that is a very git oriented term.

Clone is kind of redundant to download, but it feels like the command may need to be separate from download, and needs to convey that the user is cloning someone else's project and they can head in their own direction with it. When adding or downloading a dependency, you're kind of along for the ride with whatever they do in their repo (within the semantic version that you're locked at). Granted, you can edit modules in `node_modules` directly, and with abandonware I've done that before, but that breaks from the intended npm design. With a clone, you're not limited at all. Your mindset needs to be that you're responsible for this separate copy now.

This is all good stuff. Let me know what you think on the command/subcommand Yoda-ordering issue. I don't know that I have a strong opinion one way or the other, I'm just leaning towards simpler, more language-like commands. As soon as we've pulled the trigger on one of the command ordering options I'll make those changes. After that I've probably only got a few more hours of coding to implement cloning/downloading, and refactoring and removal of components. Then I think the CLI should be usable enough to start testing. I expect some things will break down once we start using it, but then we'll just go back and fix what's wrong and try again.

I got frustrated with all the git libraries for Rust, and was a little frustrated with the ones for Python, and so I've dropped back to using the git binary that's installed on the system. I think that using and iterating on the CLI right now is more important than nailing a "batteries included" experience. So, on Windows 10 if you haven't already, you'll need to set up [git](https://git-scm.com/download/win) and [npm](https://blog.teamtreehouse.com/install-node-js-npm-windows). Those versions should run identically to their Linux/Mac counterparts without needing Cygwin or anything like that. I'll get instructions in the readme soon-ish, but those links will probably be what I refer users to.

**Comment by m30-jrs** [comment]
> because it's like Yoda-speak compared to the way we think and speak normally

🤣  I need to sleep on it I think.  My one comment for tonight on this topic is that what I would like to find is some way to keep the tool accessible to non-programmers while still grouping related commands together....  OK, I lied, now I am wondering if all of the "commands" I listed last night are fundamentally unique compared to commands related to generation of BoM and docs.  

What I mean is, when would the sliderule cli ever "CRUD" anything but components?  I bet we could drop the "component" from everything I said last night and the result would be everything still makes sense.  In fact, it might make more sense.  

> If I understand you correctly, you're talking about a user making changes to one of their own remote components, and then pushing it back to the remote repo?

Let me post a completely separate thread about this.  (Note to self, I think the answer to my question is that the use case I am thinking of but not clearly expressing is above the scope of the cli, but may be something other levels of sliderule architecture could help with.)

> If we do that, we should probably have pseudo-commands that exist just to run several commands at once to save users from a lot of tedium.

👍 Great minds think alike.  This was something I was pondering while getting ready today but had not had time to post.

> I think what you've laid out with separating pulling the project changes from git vs pulling the remote dependency components is fine, and is something we'd need eventually anyway. 

Cool.

> I worry about limiting the levels of parent-child relationships, because it seems like you could end up with broken remote components because they're missing a critical piece somewhere down the line.

Noted.  This is definitely an area I have not fully considered myself.  My comments were more about wanting to ensure we discussed what the proper thing to do is.  There is some level of control that defaults to doing the right thing for 80% of the cases and enables enough manual control (possibly just through low level commands like calling git and npm directly?) to cover the other 20%.  Like they say in unix, easy things should be easy and hard things should be possible.

> Clone is kind of redundant to download, but it feels like the command may need to be separate from download, and needs to convey that the user is cloning someone else's project and they can head in their own direction with it.

You know, maybe I am not enough of a git user....  Don't you use clone as kind of a checkout as well to get a local working set of files from a server?  Either way, the idea of creating your own fork of a project (beyond checking out the source) is an important detail we need to be clear about in the command structure...

> This is all good stuff. 

Glad to hear that.  I was worried toward the end that I was starting to ramble...

> I've dropped back to using the git binary that's installed on the system. I think that using and iterating on the CLI right now is more important than nailing a "batteries included" experience.

Agreed.

**Comment by jmwright** [comment]
> Don't you use clone as kind of a checkout as well to get a local working set of files from a server?

If you don't have write permission to the remote repository you've cloned, then you're stuck with a one-way relationship. You can pull changes from the remote repo, but you can't push changes to it. If you make changes to your local copy and need to pull changes from the remote later, you run the risk of conflicts. I see cloning used in three contexts:
1. As part of an installation method where the user is only consuming the software, not contributing.
2. To get someone set up to be a maintainer, meaning that they'll be given write access.
3. The remote is abandonware and you're going to maintain it locally yourself (as a lib in your project), but there's little/no risk that the old maintainer will start the project back up again (no pulls and thus no conflicts).

While it is possible to do things like changing the origin to point to a new URL and then pushing there, the CLI will essentially be creating a copy of the project that's one-way only. I may have explained that wrong, but that's what I think we need to convey. If you want to have a local copy of your friend's repo in order to contribute changes back to it, you create a fork and then clone that fork. If someone tries to contribute their changes directly back to their friend's remote repo, even if they've created a branch, and they don't have write access, their changes will be stranded in what should have been a fork for their intended purpose. The problem is, what if that user didn't know they wanted to contribute changes when they grabbed a copy through the CLI, and/or don't understand that the CLI is only a one-way copy unless they become a maintainer?

**Comment by m30-jrs** [comment]
OK, I see what you mean, but I think we are debating a semantic difference that may not mean anything to someone not familiar with git (even old school svn users like me might not see the distinction you are referring to without digging around in the meaning of clone and then only if they thought it was related to git's clone).  

I think the fundamental issue here is that when you "download" from a URL (aka, we run `git clone` or possibly even `svn co` I forget how it works in terms of permissions to check out on a publicly readable server), you are not guaranteed to be able to upload.  And, it is the repo owner for that URL that gets to decide if you can upload.  Therefor, if you are not the repo owner, you are neither guaranteed the ability to upload nor the ability to set upload permissions.  

The only way to guarantee you can upload is to download from a repo you own (either by original authorship or by forking at the remote repo level).  I wonder if we could check a repo to see if the user is the owner before downloading by URL and if not prompting the user to confirm they have upload permissions to this repo before proceeding.  We could even prompt them to look at an article about forking before exiting if they don't know or don't have upload permissions.

**Comment by jmwright** [comment]
If we switch to single commands, we may want to rename the `create` command, which currently initializes a new project, to something like `init`. It's the only command that's really different than the others. After that `create`, `add`, `remove`, `refactor`, etc will all operate on components. `create` will create a new local component, whereas `add` will pull in a remote component.

Thoughts?

**Comment by jmwright** [comment]
@m30-jrs I've gone through another set of revisions on the [command listing](https://github.com/7BIndustries/sliderule-cli/blob/master/README.md#command-listing). Let me know if you think it's heading in the right direction based on our conversations.

There are three project level commands (init/update/upload), and the rest are component level commands. I think that might make add/create/remove/refactor a little confusing. I'm wondering if those commands should go back to being postfixed with `_component`.

**Comment by jmwright** [comment]
I just re-read your comments in this thread and I missed something in the command listing. `update` should be replaced with `download` and that command has the three subcommands for more granular control of what gets updated. I'll try to get that change made tomorrow.

**Comment by m30-jrs** [comment]
>If we switch to single commands, we may want to rename the create command, which currently initializes a new project, to something like init. It's the only command that's really different than the others.

Can you clarify for me what is different between create and init?  Given that a local component needs to be created with all the same structure as a "top level" component to support automatic refactoring, the only difference I can imagine would be putting the "top level" project under version control (any ignore files would need to be present in the local component, but it would already be inside a version controlled directory).

>There are three project level commands (init/update/upload), and the rest are component level commands. I think that might make add/create/remove/refactor a little confusing.

I would propose that adding/creating/removing/refactoring components are also project level commands.  They directly affect the project structure/dependencies enabling (or disabling) the use of components in BoMs and docs.  

The place where I think we start to get into a new object of interest in the command structure is when we start discussing what operations are available to generate BoMs and docs from source.  I almost wonder if we will want a second sliderule cli tool: sliderule-make (supporting things like `make bom` and `make assembly` and `make clean`)...

>update should be replaced with download...

Yup.  And having slept on it, I think `download dependencies` should go ahead and do full recurse from the CWD (users can always manage what is being updated by following best practices and pinning their dependencies to specific versions of component sources so only things they choose to rev in their dependency lists will update regardless of recursion).

**Comment by jmwright** [comment]
I've reworked the [command listing](https://github.com/7BIndustries/sliderule-cli/blob/master/README.md#command-listing) again. I feel like we're getting close to our v1.0 commands for project structure management.

For building the BoMs and docs, why not just add a `build` command to the existing CLI?
```
sliderule-cli build [all | bom | docs]
```
The only thing that's not handled in my example there is recursion. Do we always recursively build artifacts for the local components? Do we automatically build the artifacts recursively for remote components as well?

**Comment by m30-jrs** [comment]
I just reviewed the updated command listing.  I agree, we are getting super close on the project management side of the commands.  I have been really thinking about `create` command and how it relates to top level components.  Hear me out here.

One of my core beliefs in software architecture is one should minimize "side effects".  That is, operations that do one thing (like say creating a component) that sometimes also do something else (like oh say perform repo operations).  The more we have been discussing create/init, the more this has been bugging me.  Init had the side effect of creating a component and create sometimes has the effect of doing init operations.  

I would really like to structure things so that `create` is a purely component creation (top level or local sub-component) command.  Additionally, I think it is also better if we enable users to take advantage of the project structure tools without strictly requiring them to put their project into a repo (I mean they should put things in repos, but let’s not force them).  Which leads me to the following idea.

Do the init operations on the first call to `upload`.  We really only need repo support to do uploads.  `download` would still work find for a non-initialized top-level component, it would just skip the repo portion, only running the dependency downloads.  And `upload` could easily check to see if the component is in a repo before committing the source, then prompt the user for a url if needed and determine the type of repo to initialize based in the user supplied url.

What do you think?

**Comment by jmwright** [comment]
I think that sounds good. 

I suppose it could create a situation where a user creates a local component, cd's into that components directory, and then uploads it. That would put that component in a weird local/remote limbo since it wasn't refactored properly. That issue is not unique to separate create/init operations, and is more of a documentation/workflow issue though. Maybe the CLI can check to see if the parent directory is `components` when `upload` is run, and can warn the user they need to do a refactor instead of what they're doing.

**Comment by m30-jrs** [comment]
> Maybe the CLI can check to see if the parent directory is components when upload is run, and can warn the user they need to do a refactor instead of what they're doing.

I was thinking that or we could leave a .file of some kind in sub component directories so we know it is not a top level component.  Then remove the file on refactor.

> I've reworked the sample workflow and command listing for the latest revisions. The create and upload commands are the only ones that have changed in the listing. I'll try to get the changes integrated into the CLI early next week, and get some zip files on a server so they can be linked from the readme.

Those description updates look good.  I will be eagerly waiting for zip file drop.

**Comment by m30-jrs** [comment]
> I was thinking that or we could leave a .file of some kind in sub component directories so we know it is not a top level component. Then remove the file on refactor.

Scratch that... Reverse it.  When a component becomes a top-level component (first component created or when it is refactored) then we drop a .top file.  The reason to do it this way is that becoming a top level component is a one way trip.  So, we only have to remember to add the file, not add and remove it.

> How should we handle license files? Prompt the user to add one on their own?

I would like to be more explicit than that (force a choice, even if the choice is "no reuse allowed").  There is also the bit about LICENSE files vs the LICENSE field in project.json.  I think this is another place for some discussion in another thread.

**Comment by jmwright** [comment]
>  The reason to do it this way is that becoming a top level component is a one way trip.

Does the .top file become an issue when the component is included as a remote component in another project? Or is the remote component still considered a top level component?
